librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+    /// eventually computed in `from_data` and cannot be changed after that
+    pub fn compute_hash(&self) -> Result<Multihash, Error> {
+        Ok(Sha2_256::digest(&self.canonical_data()?))
+    }
+
+    /// Check that this key is allowed to sign the entity by checking that the
+    /// `by` argument is correct:
+    ///
+    /// - either the key is owned by `self`
+    /// - or it belongs to the given certifier
+    pub async fn check_key(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        match by {
+            Signatory::OwnedKey => {
+                if !self.has_key(key) {
+                    return Err(Error::KeyNotPresent(key.to_owned()));
+                }
+            },
+            Signatory::User(uri) => {
+                let user = resolver.resolve(&uri).await?;
+                if !user.has_key(key) {
+                    return Err(Error::UserKeyNotPresent(uri.to_owned(), key.to_owned()));
+                }
+            },
+        }
+        Ok(())
+    }
+
+    /// Given a private key, compute the signature of the entity canonical data
+    /// FIXME: we should check the hash instead: it is cheaper and makes also
+    /// verification way faster because we would not need to rebuild the
+    /// canonical data at every check (we can trust the hash correctness)
+    pub fn compute_signature(&self, key: &Key) -> Result<Signature, Error> {
+        Ok(key.sign(&self.canonical_data()?))
+    }
+
+    /// Given a private key, sign the current entity
+    ///
+    /// The following checks are performed:
+    ///
+    /// - the entity has not been already signed using this same key
+    /// - this key is allowed to sign the entity (using `check_key`)
+    pub async fn sign(
+        &mut self,
+        key: &Key,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        let public_key = key.public();
+        if self.signatures().contains_key(&public_key) {
+            return Err(Error::SignatureAlreadyPresent(public_key.to_owned()));
+        }
+        self.check_key(&public_key, by, resolver).await?;
+        let signature = EntitySignature {
+            by: by.to_owned(),
+            sig: self.compute_signature(key)?,
+        };
+        self.signatures.insert(public_key, signature);
+        Ok(())
+    }
+
+    /// Check that an entity signature is valid
+    pub async fn check_signature(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver).await?;
+        if signature.verify(&self.canonical_data()?, key) {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailed)
+        }
+    }
+
+    /// Check the validity of this entty (only this revision is checked)
```

Comment:
```suggestion
    /// Check the validity of this entity (only this revision is checked).
```

Reply:
:+1: to this and all subsequent typos and suggestions :-)